### PR TITLE
Add FAQ for Forbidden (DENIED) / expand FAQ / update error message

### LINKED ISSUE #1619: Better error message when UNAUTHORIZED
UNAUTHORIZED is a common failure, and the current error message is very general to set a credential helper or credentials in a build config file.

`[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:1.0.2:build (default-cli) on project hello-spring-boot: Build image failed, perhaps you should set a credential helper name with the configuration '<to><credHelper>' or set credentials for 'gcr.io' in your Maven settings: Unauthorized for ...: 401 Unauthorized
[ERROR] {"errors":[{"code":"UNAUTHORIZED","message":"You don't have the needed permissions to perform this operation, and you may have invalid credentials. To authenticate your request, follow the steps in: https://cloud.google.com/container-registry/docs/advanced-authentication"}]}`

We can have a detailed FAQ entry and put the link in the error message.

### PR DESCRIPTION
**Pull request by chanseokoh**
Fixes #1619.

Also
- Added a new FAQ for "Forbidden (DENIED)".
- Expanded the previous FAQ for "Unauthorized".
- Simplified `HelpfulSuggestions` class.

### CODE REVIEW

File: docs/faq.md [hunk 1]
```diff
@@ -344,6 +345,12 @@ See more at [Using Google Container Registry (GCR) with Minikube](https://ryanes
 
 Jib currently requires configuring your build tool to use the appropriate [Java networking properties](https://docs.oracle.com/javase/8/docs/api/java/net/doc-files/net-properties.html) (`https.proxyHost`, `https.proxyPort`, `https.proxyUser`, `https.proxyPassword`).
 
+### What should I do when the registry responds with Forbidden or DENIED?
+
+If the registry returns "403 Forbidden" or `"code":"DENIED"`, it often means Jib successfully authenticated using credentials you provided but the credentials do not have permissions to pull or push images. Make sure your account/role has the permissions to do the operation.
```
Comment (loosebazooka) [inline docs/faq.md]: "403 Forbidden" vs `"403 Forbidden"` ? since we're doing it for the next `"code":"DENIED"`

File: docs/faq.md [hunk 2]
```diff
@@ -344,6 +345,12 @@ See more at [Using Google Container Registry (GCR) with Minikube](https://ryanes
...
+
+However, depending on registry implementations, it is also possible that the registry actually meant you are not authenticated. Sometimes, it may be due to Jib picking up different credentials, credentials incorrectly configured, or typos in auth configuration. Or, it may be that your credentials work for pulling the base image but not for pushing. See [What should I do when the registry responds with UNAUTHORIZED?](#what-should-i-do-when-the-registry-responds-with-unauthorized) to ensure you have set up credentials correctly.
```
Comment (loosebazooka) [inline docs/faq.md]:
```diff
- However, depending on ...
+ Depending on ...
```

```diff
- implementations, it is also possible ... 
+ implementations it is possible ...
```

I think you can remove the whole section about wrong credentials, and just directly link to the faq after `meant you are not authenticated.` There's not much value in repeating all the reasons for unauthorized again here.

File: docs/faq.md [hunk 3]
```diff
@@ -355,8 +362,10 @@ If the registry returns "401 Unauthorized" or `"code":"UNAUTHORIZED"`, it is oft
...
       - These parameters can also be set through properties: [Maven](https://github.com/GoogleContainerTools/jib/tree/master/jib-maven-plugin#system-properties) / [Gradle](https://github.com/GoogleContainerTools/jib/tree/master/jib-gradle-plugin#system-properties)
+* There exist different auth configurations in multiple places, and Jib is not picking up the auth information you are working on.
 * You configured a credential helper, but the helper is not on `$PATH`. This is especially common when running Jib inside IDE where the IDE binary is launched directly from an OS menu and does not have access to your shell's environment.
 * Typos in username, password, image names, or registry names.
+* You are using a private registry without HTTPS. See [How can I diagnose problems pulling or pushing from remote registries?](#how-can-i-diagnose-problems-pulling-or-pushing-from-remote-registries) if this is the case.
```
Comment (loosebazooka) [inline docs/faq.md]: remove "if this is the case"

File: docs/faq.md (hunk 2 above)
Comment (TadCordle) [reply docs/faq.md]: >```diff
>- implementations, it is also possible ... 
>+ implementations it is possible ...
>```

Actually I think this comma should be here.

File: docs/faq.md [hunk 4]
```diff
@@ -355,8 +362,10 @@ If the registry returns "401 Unauthorized" or `"code":"UNAUTHORIZED"`, it is oft
       - Configuring credential helpers: [`<from/to><credHelper>`](https://github.com/GoogleContainerTools/jib/tree/master/jib-maven-plugin#using-docker-credential-helpers) for Maven / [`from/to.credHelper`](https://github.com/GoogleContainerTools/jib/tree/master/jib-gradle-plugin#using-docker-credential-helpers) for Gradle
       - Specific credentials (not recommend): [`<from/to><auth><username>/<password>`](https://github.com/GoogleContainerTools/jib/tree/master/jib-maven-plugin#using-specific-credentials) or in [`settings.xml`](https://github.com/GoogleContainerTools/jib/tree/master/jib-maven-plugin#using-maven-settings) for Maven / [`from/to.auth.username/password`](https://github.com/GoogleContainerTools/jib/tree/master/jib-gradle-plugin#using-specific-credentials) for Gradle
       - These parameters can also be set through properties: [Maven](https://github.com/GoogleContainerTools/jib/tree/master/jib-maven-plugin#system-properties) / [Gradle](https://github.com/GoogleContainerTools/jib/tree/master/jib-gradle-plugin#system-properties)
+* There exist different auth configurations in multiple places, and Jib is not picking up the auth information you are working on.
```
Comment (TadCordle) [inline docs/faq.md]: `Different auth configurations exist in multiple places, ...`

File: jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestions.java [hunk 5]
```diff
@@ -193,27 +172,37 @@ public String forCacheDirectoryNotOwned(Path cacheDirectory) {
...
-      return forNoCredentialHelpersDefined(
-          baseImageCredHelperConfiguration, baseImageAuthConfiguration.apply(registry));
-    }
-    if (noCredentialsDefinedForTargetImage
-        && registry.equals(targetImageReference.getRegistry())
-        && repository.equals(targetImageReference.getRepository())) {
-      return forNoCredentialHelpersDefined(
-          targetImageCredHelperConfiguration, targetImageAuthConfiguration.apply(registry));
+
+    final String unauthorizedEntity;
```
Comment (TadCordle) [inline jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestions.java]: Maybe we can just do `String unauthorizedEntity = registry;` and remove the else clause so we can declare and initalize on the same line.

File: docs/faq.md (hunk 2 above)
Comment (chanseokoh) [reply docs/faq.md]: Retaining the comma.

**Comment by chanseokoh** [comment]
Thanks for wording suggestions.

**Comment by chanseokoh** [comment]
> I would like to see us move away from the `HelpfulSuggestions` model in the future

Agreed. If you have some vague idea, any, I'm curious to hear later.

**Pull request merged**